cpp/src/io/parquet/page_enc.cu

```diff
@@ -131,38 +166,45 @@ __global__ void __launch_bounds__(512) gpuInitPageFragments(PageFragment *frag,
     dtype_len_in = (dtype == BYTE_ARRAY) ? sizeof(nvstrdesc_s) : dtype_len;
   }
   __syncthreads();
-  nrows = s->frag.num_rows;
-  for (uint32_t i = 0; i < nrows; i += 512) {
+
+  nvals           = s->frag.num_leaf_values;
+  start_value_idx = s->start_value_idx;
+
+  for (uint32_t i = 0; i < nvals; i += 512) {
     const uint32_t *valid = s->col.valid_map_base;
-    uint32_t row          = start_row + i + t;
-    uint32_t is_valid     = (i + t < nrows && row < s->col.num_rows)
-                          ? (valid) ? (valid[row >> 5] >> (row & 0x1f)) & 1 : 1
+    uint32_t val_idx      = start_value_idx + i + t;
+    uint32_t is_valid     = (i + t < nvals && val_idx < s->col.num_values)
+                          ? (valid) ? (valid[val_idx >> 5] >> (val_idx & 0x1f)) & 1 : 1
```

Comment:
As a general comment, I'm more of a fan of using / instead of >> to do this kind of thing.  cuio in general seems to prefer using shift, but it's kind of an archaic way of doing it.  Might be worth considering making this a style change as part of the refactor.

Reply:
In defense of logical operators, it's been my experience that many people tend to forget the underlying assumption that arithmetic operators for power-of-two modulo/divisions are only equivalent to the logical operators for unsigned types. This is demonstrably true in libcudf, where signed arithmetic is encouraged by cudf::size_type, which results in suboptimal code generation in the majority of cases (though not a problem where style is valued over substance).

Reply:
It's my experience that people who use explicit logical operators don't realize that the most base optimizing compiler will already make this optimization for you: https://godbolt.org/z/Y4G44d

Reply:
> It's my experience that people who use explicit logical operators for division by a power of 2 don't realize that the most basic optimizing compiler will already make this optimization for you: https://godbolt.org/z/Y4G44d

https://godbolt.org/z/9KfoTh

"In all affairs it's a healthy thing now and then to hang a question mark on the things you have long taken for granted." -- Bertrand Russell

Reply:
> Which proves my point about people forgetting the underlying assumption about unsigned math.
https://godbolt.org/z/9KfoTh

Your example proves that you should always use `/` instead of trying to be clever and use bitshifts as you might forget your dividend is signed and using bitshifts could yield incorrect results.  In the case of unsigned, the compiler will already do the optimization for you. In the case of signed, the compiler will still generate more optimal code when it knows the divisor is a power of 2 than if the divisor was only known at runtime.

Reply:
> Your example proves that you should always use `/` instead of trying to be clever and use bitshifts as you might forget your dividend is signed and using bitshifts could yield incorrect results.

No. If anything, my example demonstrates that - in the words of Karl Popper - demands for conformity, in the end, are demands for mediocrity.

Reply:
> No. If anything, my example demonstrates that - in the words of Karl Popper - demands for conformity, in the end, are demands for mediocrity.

If you want to bring Karl Popper into this---you've put forth a hypothesis and it has been demonstrably falsified. What's left?

Reply:
> If you want to bring Karl Popper into this---you've put forth a hypothesis and it has been demonstrably falsified. What's left?
I believe that you believe that it has been demonstrably falsified, but I think you should re-read my initial comment, let me re-state it here: "many people tend to forget the underlying assumption that arithmetic operators for power-of-two modulo/divisions are only equivalent to the logical operators for unsigned types".
1. The godbolt snippets demonstrate the code generation part to be true
2. The fact that you responded with a godbolt link to code referring to unsigned types and apparently generalized this as a reason to always use arithmetic operators without considering sign-ness would be evidence of the "people tend to forget" bit
I can also make the verifiable prediction that replacing all occurences of arithmetic modulo-power-of-2 in libcudf would result in smaller, more efficient generated code, while replacing the logical operators by arithmetic ones would have the opposite effect (Not that I'm suggesting to do that).

Reply:
> I can also make the verifiable prediction that replacing all occurences of arithmetic modulo-power-of-2 in libcudf would result in smaller, more efficient generated code

Sure, but I'll include my own prediction that it would also increase the number of actual and potential of incredibly difficult to triage bugs---something cuIO code seems to have suffered from historically.

Reply:
> Sure, but I'll include my own prediction that it would also increase the number of actual and potential of incredibly difficult to triage bugs---something cuIO code seems to have suffered from historically.

Even assuming this is a fact and not an opinion, I doubt any cuIO bug (regardless of triaging difficulty) has ever been due to using a logical operator where an arithmetic operator would have had the same result. One could in fact easily argue the opposite to be true, since the use of the modulo operator get the bit position within a uint32 or a lane id would imply a logical bit extraction (the expected result of -1 % 32 would be 31 and not -1 here) rather than an arithmetic operation.
Also, I believe this would be an unverifiable prediction, and probably [moving the goal posts](https://en.wikipedia.org/wiki/Moving_the_goalposts).